src/app/CommandSender.h

```diff
@@ -114,6 +114,62 @@ class CommandSender final : public Messaging::ExchangeDelegate
         virtual void OnDone(CommandSender * apCommandSender) = 0;
     };
 
+    // CommandSender methods are called outside of SDK. By having parameters passed using this
+    // struct we centralize defaults within the struct instead of the function's declaration
+    // parameters list.
+    struct ConfigParameters
+    {
+        ConfigParameters & SetRemoteMaxPathsPerInvoke(uint16_t aRemoteMaxPathsPerInvoke)
+        {
+            mRemoteMaxPathsPerInvoke = aRemoteMaxPathsPerInvoke;
+            return *this;
+        }
+
+        // If mRemoteMaxPathsPerInvoke is 1, this will allow the CommandSender client to contain only one command and
+        // doesn't enforce other batch commands requirements.
+        uint16_t mRemoteMaxPathsPerInvoke = 1;
+    };
+
+    // CommandSender methods are called outside of SDK. By having parameters passed using this
+    // struct we centralize defaults within the struct instead of the function's declaration
+    // parameters list.
+    struct AdditionalCommandParameters
+    {
+        // gcc bug requires us to have the constructor below
+        // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96645
+        AdditionalCommandParameters() {}
+
+        AdditionalCommandParameters & SetStartOrEndDataStruct(bool aStartOrEndDataStruct)
+        {
+            mStartOrEndDataStruct = aStartOrEndDataStruct;
+            return *this;
+        }
+
+        Optional<uint16_t> mCommandRef;
```

Comment:
This is pretty annoying in the sense that we are requiring consumers to provide a command ref and maintain command refs themselves.

It might make more sense for the CommandSender to allocate the CommandRefs itself, and have the AdditionalCommandParameters be a mutable inout param on "prepare".

Followup for this part is fine.

Reply:
My only problem with this is we then need to maintain a list of Sent command on behalf of the caller. I think allowing that managing logic where it tracks the outbound calls reduce the burden on `CommandSender` so we don't end up with overhead for the non batch case like we have in `CommandHandler` (which is unavoidable there)

Reply:
Why do we need to maintain a list of sent commands?  We're already tracking our count of encoded commands, right?  Can we not just use that for the CommandRef?

But again: followup to figure this out is probably best.

Reply:
The answer I would suggest in that case is "propagate the command ref back to the python code".

If we leave it to API consumers, it leaves more room for creating invalid requests (e.g. repeated command refs; not sure whether there are any other constraints on them)....

Reply:
Okay I will implement that behaviour now and update API docs to reflect that. I realized that you responded to this after my most recent PR push